rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,713 @@
+to whatever attributes we want to be able to interrogate, and so this provides an extremely powerful
+primitive to be able to take advantage of existing code, as well as support new attributes from
+other crates that get utilized in the future.
+
+Additionally, the ability to specify custom metadata using the `configurable` attribute is a
+powerful escape hatch when we need to encode behaviours, or inline relevant data, to configuration
+types and fields that aren't related to the schema of a Vector configuration itself... or aren't
+_possible_ to encode in JSONSchema. For example, this could be used for something simple, like in
+the above example, where we're defining the status of the sink implementation as beta. This might be
+used to drive the generated content of the vector.dev website and documentation.
+
+Some constraints might be harder to express in a JSONSchema, however, such as whether or not a sink
+supports acknowledgements. Whether or not a sink supports acknowledgements isn't terribly relevant
+to the Vector configuration itself, beyond validating whether or not any fields which toggle it on
+or off have been set right, and so on. There's a semantic relevance, however, is thatg knowing a
+sink does or doesn't support acknowledgements could allow a validator to surface issues to user.
+For example, if they have acknowledgements enabled on a sink but their source does not support
+acknowledgements, then end-to-end acknowledgements would not be able to actually function. While the
+configuration loading code can also detect this, being able to provide these semantic definitions
+within the schema itself allows us to more generically encode these types of behaviors and allow
+external tools, which don't have the benefit of running Vector directly, to correctly suss out
+incompatibilities and misconfigurations.
+
+Further, and following the example code itself, this can allow us to enforce constraints that are
+only partially able to be represented in JSONSchema, such as aliased fields. While JSONSchema
+already supports the ability to define a schema such that a field can be represented by multiple
+names, it has no concept of a deprecated field. Utilizing custom metadata, we can encode metadata
+that indicates which field name variant is the deprecated one. This can be used not only to drive
+behavior in the generated documentation, but in other tools as well, such as automatically
+transforming one version of a configuration to a newer version by analyzing the schemas, and being
+able to reason that if field X used to be able to be referenced via A or B, and now only B is
+allowed, we can just rename A to B. We could also check that schemas don't remove fields unless
+those fields were already marked as deprecated, being able to enforce our unofficial guideline of
+not removing fields unless they've been marked as deprecated for at least one release.
+
+#### 4. Utilize `schemars` to generate the actual JSONSchema
```

Comment:
I think I'm failing to see, still, what the JSONSchema output would look like. Would it be possible to add an example based on the example `Configurable` above?

Reply:
Apologies, I see you included that above as a note to self 😄 I am definitely interested in seeing the output here.